README.md
# Ecom API

* [Configuration](#configuration)
  * [Environment Variable](#environment-variables)
    * [App](#env-app)
    * [Google](#env-google)
    * [Stripe](#env-stripe)
    * [Postgres](#env-postgres)
  * [File Structure](#file-structure)
* [Deployment](#deployment)
  * [Google Compute Engine](#gce)
* [App Architecture](#arch)
  * [Model](#arch-model)
  * [Service](#arch-service)
  * [App](#arch-app)
* [API](#api)
  * [CreateCart](#CreateCart)
  * [AddItemToCart](#AddItemToCart)
  * [GenerateCustomerDevKey](#GenerateCustomerDevKey)
  * [SignInWithDevKey](#SignInWithDevKey)
  * [PlaceOrder](#PlaceOrder)
  * [StripeCheckout](#StripeCheckout)

## <a name="configuration"></a>Configuration

The `ecom-api` executable accepts configuration through the use of environment variables.


### <a name="environment-variables"></a>Environment Variables

Configuration is split into four groups; [App](#env-app), [Google](#env-google), [Stripe](#env-stripe) and [PostgreSQL](#env-postgres) prefixed with `ECOM_APP_`, `ECOM_GOOGLE_`, `ECOM_STRIPE_` and `ECOM_PG_` respectively. The Stripe configuration is optional and is only require if you intend to take payments with the Stripe payment gateway.

The Google configuration includes Firebase Auth and is mandatory since the API requires Firebase Auth for Authentication and Authorisation.


#### <a name="env-app"></a>App

| Env Var                  | Required | Default | Description |
| -------------            | -------- | ------- | ------------|
| **`PORT`**                   | Optional | 8080    | Usually set by the container. For example, Google App Engine sets this on your behalf. |
| **`ECOM_APP_TLS_MODE`**      | Optional | disable | To enable set the value to `enable`. If enabled you must set `ECOM_APP_TLS_CERT` and `ECOM_APP_TLS_KEY`. |
| **`ECOM_APP_TLS_CERT`**      | Depends  |         | Path to TLS certificate. e.g. `/etc/secret-volume/tls/api_spycameracctv_com/cert.pem`. Required if `ECOM_APP_TLS_MODE=enable`. |
| **`ECOM_APP_TLS_KEY`**       | Depends  |         | Path to TSL certificate key. e.g. `/etc/secret-volume/tls/api_spycameracctv_com/key.pem` |
| **`ECOM_APP_ROOT_EMAIL`**   | Optional |         |
| **`ECOM_APP_ROOT_PASSWORD`** | Required |         |
| **`ECOM_APP_ENABLE_STACKDRIVER_LOGGING`** | Optional | on | Accepts a value of `on` or `off` to switch the stack driver JSON formatted logging. |
| **`ECOM_APP_ENDPOINT`** | Required | | An absolute and secure URL endpoint to the API Service. Example URL https://c90e3367.ngrok.iolocalhost:8080. |


#### <a name="env-google"></a>Google

| Env Var                         | Required | Default | Description |
| -------------                   | -------- | ------- | ------------|
| **`ECOM_GAE_PROJECT_ID`**       | Required |         | Set the value to the Google Project ID where the GAE App is running. For example, `open247-gae`. |
| **`ECOM_FIREBASE_PUBLIC_CONFIG`**  | Required |         | base64 encoded string firebase config JSON string. |
| **`ECOM_FIREBASE_PRIVATE_CREDENTIALS`** | Required |         | Use either the filepath of the Firebase Service Account Credentials file or provide a Base64 encoded string. e.g. `/etc/secret-volume/service_account_credentials/ecom-test-fa3e406ce4fe.json` (or base64 encoded JSON string) |
| **`ECOM_GOOGLE_PUBSUB_PUSH_TOKEN`** | Required | A secret token used for basic auth to the push endpoint. |

#### <a name="env-stripe"></a>Stripe

The stripe settings are optional and only required if you are processing payments with the Stripe checkout. If you do not use Stripe, you do not need to pass either of the following environment variables to the ecom service. If you are using Stripe, make sure you pass both the secret key and signing key. The secret key is used for payments and the signing key is used to verify confirmations from Stripe webhooks.

| Env Var                       | Required | Default | Description |
| -------                       | -------- | ------- | ----------- |
| **`ECOM_STRIPE_SECRET_KEY`**  | Depends  |         | Found in the [API Keys](https://dashboard.stripe.com/test/apikeys) section of the Stripe Dashboard. |
| **`ECOM_STRIPE_SIGNING_KEY`** | Depends  |         | Found in the [Webhooks](https://dashboard.stripe.com/test/webhooks) section of the Stripe Dashboard. |
| **`ECOM_STRIPE_SUCCESS_URL`** | Depends  |         | URL to redirect to when checkout payment is successful. |
| **`ECOM_STRIPE_CANCEL_URL`**  | Depends  |         | URL to redirect to when checkout payment is cancelled. |

#### <a name="env-postgres"></a>Postgres

| Env Var                   | Required | Default | Description |
| -------------             | -------- | ------- | ------------|
| **`ECOM_PG_HOST`**            | Required |         | Example `35.234.153.166` |
| **`ECOM_PG_PORT`**            | optional | 5432	   | Postgres TCP/IP Port |
| **`ECOM_PG_DATABASE`**        | Required |         | Postgres database name. Example `ecom_dev` |
| **`ECOM_PG_USER`**            | Required |         | Postgres database user. Example `postgres` |
| **`ECOM_PG_PASSWORD`**        | Required |         | Postgres database password. |
| **`ECOM_PG_SSLMODE`**         | Optional | disable | Example `verify-ca`. |
| **`ECOM_PG_SSLCERT`**         | Optional |         | e.g. `/etc/secret-volume/pg/client-cert.pem`. |
| **`ECOM_PG_SSLKEY`**          | Optioanl |         | Example `/etc/secret-volume/pg/client-key.pem`. |
| **`ECOM_PG_SSLROOTCERT`**     | Optional |         | Example `/etc/secret-volume/pg/server-ca.pem`. |
| **`ECOM_PG_CONNECT_TIMEOUT`** | Optional | 10      | Postgres connection timeout in seconds. Example `5`. |


`ECOM_GOOGLE_PROJECT_ID` can be obtained from the Firebase Console. When you create a Firebase app, Google creates a Google Cloud project with the same project ID. The Firebase Auth is accessible via the Firebase control panel and is backed by a project of the same ID in the Google Cloud Console.

`ECOM_FIREBASE_PRIVATE_CREDENTIALS` is the service account key file found in the Google Cloud Console, or Firebase console under the service account keys section.
`ECOM_PG_PORT` should use a private IP if running on GKE or GCE.

GAE sets `PORT` for each container it starts. Hard coding the port to 8080 causes an error message to appear in Stackdriver about nginx reverse proxy misconfiguration, resulting in poor performance.

### <a name="file-structure"></a>File Structure

If you opt to host the `ecom-api` using a virtual machine instance, and opt to use private SSL certificates, the `ecom-api` executable must have access to its own private disk mounted at `/etc/secret-volume`. This directory contains three directories `pg`, `service_account_credentials` and tls housing the PostgreSQL key files for SSL connections, Firebase Service Account files and SSL certificates respectively.

```
secret-volume/
├── pg
│   ├── client-cert.pem
│   ├── client-key.pem
│   └── server-ca.pem
├── service_account_credentials
│   ├── ecom-test-fa3e406ce4fe.json
│   └── test-spycameracctv-firebase-adminsdk-b06ml-46cd9030e2.json
└── tls
    ├── api_spycameracctv_com
    │   ├── cert.pem
    │   └── key.pem
    └── star_open24seven_co_uk
        ├── cert.pem
        └── key.pem
```


## <a name="deployment"></a>Deployment

### <a name="gce"></a>Google Compute Engine
GAE offers free SSL endpoints and provides a proxy method to securely connect to Postgres, so there is no need to deploy the secret volume containing either the `secrets/tls` and `secrets/pg` directories. The `ecom-api` executable accepts the service account credentials as a Base64 encoded string passed using the `ECOM_FIREBASE_PRIVATE_CREDENTIALS` so there is often no need to package the service credentials file. The app can be configured entirely using environment variables.

See Google's documentation for [Deploying Containers on VMs and Managed Instance Groups](https://cloud.google.com/compute/docs/containers/deploying-containers).

Go 1.12 Standard Environment for App Engine recommends logging to stdout and stderr. See [Writing Application Logs](https://cloud.google.com/appengine/docs/standard/go112/writing-application-logs).

Other blog articles indicate that calling the Google Cloud API has higher latency than streaming to stdout from the service. This is better for cross platform compatibility as the Google Cloud Go library will lock to GCP only.

Container terminated by the container manager on signal 9.

`SIGKIL`

Appears that messages sent to stdout or stderr must have a terminating newline character to be seen in the Log Viewer.


``` go
fmt.Fprintf(os.Stdout, "Test message\n")
```

In your application code, look for the `X-Cloud-Trace-Context` HTTP header of incoming requests. Extract the trace identifier from the header.

Set the trace identifier in the LogEntry trace field of your app log entries. The expected format `isprojects/[PROJECT_ID]/traces/[TRACE_ID]`.

```
"X-Cloud-Trace-Context: TRACE_ID/SPAN_ID;o=TRACE_TRUE"
```

Where:
+ **`TRACE_ID`** is a 32-character hexadecimal value representing a 128-bit number. It should be unique between your requests, unless you intentionally want to bundle the requests together. You can use UUIDs.

+ **`SPAN_ID`** is the decimal representation of the (unsigned) span ID. It should be 0 for the first span in your trace. For subsequent requests, set `SPAN_ID` to the span ID of the parent request. See the description of `TraceSpan(REST, RPC)` for more information about nested traces.

+ **`TRACE_TRUE`** must be 1 to trace this request. Specify 0 to not trace the request.


```
$ gcloud compute instances list
```

#### <a name="updating-a-container"></a>Updating a container

Update container declaration on the the instance
Stops and restarts the instance to actuate

```
$ gcloud compute instances update-container test-spy --container-image gcr.io/cloud-marketplace/google/nginx1:1.13
```

#### Connecting to a container using SSH

```
cloud compute ssh --project [PROJECT_ID] --zone [ZONE] [INSTANCE_NAME]
```


Where:
+ `[PROJECT_ID]` is the ID of the project that contains the instance.
+ `[ZONE]` is the name of the zone in which the instance is located.
+ `[INSTANCE_NAME]` is the name of the instance.

```
$ gcloud compute ssh --project spycameracctv-d48ac --zone europe-west2-b spy-test-instance
```

Reserving a static external IP address

```
$ gcloud compute instances list
```


#### Create a new static IP address

```
$ gcloud compute addresses create test-spycameracctv-ip

For the following address:
 - [test-spycameracctv-ip]
choose a region or global:
 [1] global
 [2] region: asia-east1
 [3] region: asia-east2
 [4] region: asia-northeast1
 [5] region: asia-south1
 [6] region: asia-southeast1
 [7] region: australia-southeast1
 [8] region: europe-north1
 [9] region: europe-west1
 [10] region: europe-west2
 [11] region: europe-west3
 [12] region: europe-west4
 [13] region: northamerica-northeast1
 [14] region: southamerica-east1
 [15] region: us-central1
 [16] region: us-east1
 [17] region: us-east4
 [18] region: us-west1

CHANGELOG.md
# CHANGELOG
## v0.64.0 (Wed, 11 Dec 2019)
+ Stripe checkout and order handling publish events.

## v0.63.0 (Wed, 27 Nov 2019)
+ `OpPlaceOrder`, `OpGetOrder` and `OpListOrders`.

## v0.62.3 (Wed, 13 Nov 2019)
+ Validate uuid on path parameters for HTTP DELETE /developer-keys/:id
+ Fix issue with stackdriver logging not enabling.
+ Promo rules return `promo_rule_code` attribute in JSON responses.

## v0.62.2 (Tue, 5 Nov 2019)
+ Fix request validation for `PATCH /address` operation `OpUpdateAddress`.
+ Improved logging in main and app level handlers for coupons and addresses.
+ `PromoRuleCode` attribute added to coupons.
+ Minor changes to Open API Def.
+ Update deps using `go get -v all`.

## v0.62.1 (Sun, 3 Nov 2019)
+ Fix sysinfo JSON response to include Stripe and endpoint.

## v0.62.0 (Sat, 2 Nov 2019)
+ Add Stackdriver Profiler when
+ Fix issue with missing `created` and `modified` values when creating a new promo rule.
+ Alter compiler options to use -gcflags=all='-N -l' for compute engine.
+ Update all dependencies using `go get -v all`.

## v0.61.4 (Thu, 31 Oct 2019)
+ Fixes issues with JSON responses for promo_rule objects.
+ [product_id attribute missing from HTTP GET /promo-rules/{id} response](https://bitbucket.org/andyfusniakteam/ecom-api-go/issues/20/product_id-attribute-missing-from-http-get)
+ [HTTP POST /promo-rules returns a blank promo_rule_code attribute](https://bitbucket.org/andyfusniakteam/ecom-api-go/issues/23/http-post-promo-rules-returns-a-blank)

## v0.61.3 (Tue, 22 Oct 2019)
+ Add site to the Open API Def YAML.
+ Ignore google*.html in git.
+ Tidy code

## v0.61.2 (Tue, 22 Oct 2019)
+ Return structured JSON for 401 Unauthorized and 403 Forbidden responses.

## v0.61.1 (Tue, 22 Oct 2019)
+ Fix authorization issues.
+ Remove Google vertification handler and replace with dedicated HTTP service `google-verifcation`.
+ Code tidy `app/create_webhook.go` and fix logging of pointer values.

## v0.61.0 (Mon, 21 Oct 2019)
+ Google Pub/Sub broadcasting of messages to handle HTTP POSTs to webhook endpoints.
+ `overselling` for inventory.
+ `ErrCodeWebhookPostFailed` error handling for HTTP POST failures.
+ Most handlers use pointers for the request body to detect missing attributes.
+ Remove nesting from Go error handling.
+ `ECOM_APP_ENDPOINT env var introduced.
+ `ECOM_GOOGLE_PUBSUB_PUSH_TOKEN` secret token for basic auth to the push endpoint.
+ Initial creating of topics and subscriptions to handle webhooks.
+ `/privte-pubsub-events` and `/privte-pubsub-broadcast` resources.
+ Webhooks calls implemented.
+ Makefile has two new deployments for test and live.
+ GAE now supports go113 so go.mod file is updated to use Go 1.13.3.
+ Open API Def file improved to near match API Service.
+ `offer_id` added to `price` table.


## v0.60.1 (Tue 1 Oct 2019)
+ Minor fixes to schema (`shipping_code` on `shipping_tariff` table made non unique)
+ Remove verbosity of schema create, drop and load shell scripts.
+ `OpGetPromoRule` GetPromoRuleHandler validation fix for URL param.
+ `OpGetPromoRule` fix 404 response for promo rule not found.

## v0.60.0 (Tue 1 Oct 2019)
+ Webhooks CRUDL for configuration only.
+ Google Pub/sub connection on startup.
+ Product to product group associations impemented.
+ Product to product associations implemented.
+ Coupons in the cart implemented.
+ Inventory implemented.
+ Shipping Tariffs implemented.
+ `offer_price` added to `price` table in schema.
+ Offer activation.
+ Cart products fixes and Open API def updates.
+ Shipping Tariffs spelling change.

## v0.58.1 (Fri 13 Sep 2019)
+ Fixes issues with the Postgres table creation script order in which the tables are created.

## v0.58.0 (Fri 13 Sep 2019)
+ ecom binary `ECOM_FIREBASE_PROJECT_ID` deprecated.
+ ecom binary `ECOM_FIREBASE_WEB_API_KEY` and `ECOM_FIREBASE_CREDENTIALS` become `ECOM_FIREBASE_PUBLIC_CONFIG` and `ECOM_FIREBASE_PRIVATE_CREDENTIALS` respectively.
+ README.md removes all documenation (held in separate repo and in the Open API Spec Definition).
+ `OpActivateOffer`
+ New handlers for product to category relations.
+ New handlers for adding products to carts.
+ Implements Shipping Tariffs.
+ Implements Price Lists.
+ Implements Coupons.
+ Implements Promo Rules.
+ Implements Inventory.
+ Implements Offers.
+ Implements Carts Products.
+ Improve Developer Keys
+ Implements Addresses.
+ Implements Prices.
+ `/users` has pagination skeleton but not yet working.
+ New `/products-categories` resource.
+ Implements `/products?include=prices,images` although it's not determined if it will be used in the SDK.
+ Merged `/admin` resource to universal `/users` resource.
+ `categories-tree` handles hierarchial data. `categories` handles flat nested set lists.
+ `clientError` and `serverError` functions introduced to remove boilerplate code.
+ Context values use named types to prevent golint errors.
+ Firebase Config is returned in the `/config` resource so the client doesn't need to hard code it.
+ Fixes CORS issue with `PUT` not being allowed.
+ Fixes golint warnings.
+ Postgres Schema changes.
+ Open API Definition file is not up to date with this release.

## v0.57.0 (Mon 5 Aug 2019)
+ Cart split into `carts` and `cart_items` tables.
+ Fix CORS bug with public routes.
+ Updated Open API Spec.

## v0.56.0 (Thu 1 Aug 2019)
+ Fix authorization on `OpDeleteCustomerDevKey`.
+ Apply `"object": "image"` to `OpGetImage` and `OpUpdateTierPricing`.
+ Improve OpenAPI `api.yaml` contract.

## v0.55.0 (Tue 30 Jul 2019)

cmd/app/main.go
package main

import (
	"context"
	"database/sql"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"time"

	"bitbucket.org/andyfusniakteam/ecom-api-go"
	model "bitbucket.org/andyfusniakteam/ecom-api-go/model/postgres"
	service "bitbucket.org/andyfusniakteam/ecom-api-go/service/firebase"
	"cloud.google.com/go/pubsub"
	"firebase.google.com/go"
	"firebase.google.com/go/auth"
	"github.com/go-chi/chi"
	_ "github.com/lib/pq"
	"github.com/rs/cors"
	log "github.com/sirupsen/logrus"
	"google.golang.org/api/option"
)

// set at compile-time using -ldflags "-X main.version=$VERSION"
var version string

const maxDbConnectAttempts = 3

var (
	//
	// PostgreSQL Database settings
	//

	// Name of host to connect to.
	pghost = os.Getenv("ECOM_PG_HOST")

	// Port number to connect to at the server host, or socket file name extension
	// for Unix-domain connections.
	pgport = os.Getenv("ECOM_PG_PORT")

	// The database name. Defaults to be the same as the user name.
	pgdatabase = os.Getenv("ECOM_PG_DATABASE")

	// PostgreSQL user name to connect as. Defaults to be the same as the operating system name
	// of the user running the application.
	pguser = os.Getenv("ECOM_PG_USER")

	// Password to be used if the server demands password authentication.
	pgpassword = os.Getenv("ECOM_PG_PASSWORD")

	// This option determines whether or with what priority a secure SSL TCP/IP connection will be
	// negotiated with the server. There are six modes:
	// 	disable
	// 		only try a non-SSL connection
	//
	// 	allow
	//		first try a non-SSL connection; if that fails, try an SSL connection
	//
	// 	prefer (default)
	//		first try an SSL connection; if that fails, try a non-SSL connection
	//
	//	require
	//		only try an SSL connection. If a root CA file is present, verify the
	//		certificate in the same way as if verify-ca was specified
	//
	// 	verify-ca
	//		only try an SSL connection, and verify that the server certificate
	//		is issued by a trusted certificate authority (CA)
	//
	// 	verify-full
	//		only try an SSL connection, verify that the server certificate is issued by
	//		a trusted CA and that the server host name matches that in the certificate
	pgsslmode = os.Getenv("ECOM_PG_SSLMODE")

	// This parameter specifies the file name of the client SSL certificate, replacing the default
	// ~/.postgresql/postgresql.crt. This parameter is ignored if an SSL connection is not made.
	pgsslcert = os.Getenv("ECOM_PG_SSLCERT")

	// This parameter specifies the location for the secret key used for the client certificate. It can
	// either specify a file name that will be used instead of the default ~/.postgresql/postgresql.key, or
	// it can specify a key obtained from an external "engine" (engines are OpenSSL loadable modules). An
	// external engine specification should consist of a colon-separated engine name and an engine-specific
	// key identifier. This parameter is ignored if an SSL connection is not made.
	pgsslkey = os.Getenv("ECOM_PG_SSLKEY")

	// This parameter specifies the name of a file containing SSL certificate authority (CA)
	// certificate(s). If the file exists, the server's certificate will be verified to be
	// signed by one of these authorities.
	pgsslrootcert = os.Getenv("ECOM_PG_SSLROOTCERT")

	// Maximum wait for connection, in seconds (write as a decimal integer string). Zero or
	// not specified means wait indefinitely. It is not recommended to use a timeout of
	// less than 2 seconds.
	pgconnectTimeout = os.Getenv("ECOM_PG_CONNECT_TIMEOUT")

	//
	// Google settings
	//
	projectID       = os.Getenv("ECOM_GOOGLE_PROJECT_ID")
	credentialsFile = os.Getenv("ECOM_GOOGLE_CREDENTIALS")

	//
	// Application settings
	//
	port        = os.Getenv("ECOM_APP_PORT")
	tlsModeFlag = os.Getenv("ECOM_APP_TLS_MODE")
	tlsCertFile = os.Getenv("ECOM_APP_TLS_CERT")
	tlsKeyFile  = os.Getenv("ECOM_APP_TLS_KEY")
)

const (
	// secretVolume points to the base path for the mounted drive or secret files using k8s or docker mount
	secretVolume = "/etc/secret-volume"

	// directory in the secret volume that holds all Service Account Credentials files
	sacDir      = "service_account_credentials"
	pubsubTopic = "ecom-api"
)

func initLogging() {
	// Output logs with colour
	log.SetFormatter(&log.TextFormatter{
		ForceColors: true,
	})

	// Output to stdout instead of the default stderr
	log.SetOutput(os.Stdout)

	// Log debug level severity or above.
	log.SetLevel(log.DebugLevel)
}

func healthCheckHandler(w http.ResponseWriter, r *http.Request) {
	w.WriteHeader(200)
	w.Write([]byte("ok"))
}

func infoHandler(w http.ResponseWriter, r *http.Request) {
	w.WriteHeader(http.StatusOK) // 200 OK
	fmt.Fprintf(w, fmt.Sprintf(`{"app_version":"%s"}`, version))
}

func exists(path string) (bool, error) {
	_, err := os.Stat(path)
	if err == nil {
		return true, nil
	}
	if os.IsNotExist(err) {
		return false, nil
	}
	return true, err
}

func mustHaveFile(path, title string) {
	ex, err := exists(path)
	if err != nil {
		log.Fatalf("failed to determine if %s %s exists: %v", title, path, err)
	}
	if !ex {
		log.Fatalf("cannot find %s %s. Check permissions.", title, path)
	}
	log.Infof("%s: %s", title, path)
}

func main() {
	initLogging()

	log.Infof("app version %s started", version)

	// 1. Data Source Name
	// dsn is the Data Source name. For PostgreSQL the format is "host=localhost port=5432 user=postgres password=secret dbname=mydatabase sslmode=disable". The sslmode is optional.
	if pghost == "" {
		log.Fatal("postgres host not set. Use ECOM_PG_HOST")
	}

	if pgport == "" {
		log.Info("using default port=5432 for postgres because ECOM_PG_PORT is not set")
		pgport = "5432"
	}

	if pguser == "" {
		log.Info("using default user=postgres because ECOM_PG_USER is not set")
		pguser = "postgres"
	}

	if pgdatabase == "" {
		log.Fatal("ECOM_PG_DATABASE not set.")
	}

	if pgpassword == "" {
		log.Fatal("ECOM_PG_PASSWORD not set. You must set a password")
	}

	if pgsslmode == "" {
		if pgsslkey != "" || pgsslrootcert != "" || pgsslcert != "" {
			log.Fatal("ECOM_PG_SSLMODE is not set, but one or more of ECOM_PG_SSLCERT, ECOM_PG_SSLKEY, ECOM_PG_SSLROOTCERT environment variables were set implying you intended to connect to postgres securely?")
		}
		log.Infof("using postgres sslmode=disable because ECOM_PG_SSLMODE is not set")
		pgsslmode = "disable"
	}

	if pgconnectTimeout == "" {
		log.Infof("using postgres connect_timeout=10 because ECOM_PG_CONNECT_TIMEOUT is not set")
		pgconnectTimeout = "10"
	}

	var dsn string
	if pgsslmode == "disable" {
		log.Infof("postgres running with sslmode=disable")
		dsn = fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=%s sslmode=%s connect_timeout=%s", pghost, pgport, pguser, pgpassword, pgdatabase, pgsslmode, pgconnectTimeout)
		log.Infof("postgres dsn: host=%s port=%s user=%s password=**** dbname=%s sslmode=%s connect_timeout=%s", pghost, pgport, pguser, pgdatabase, pgsslmode, pgconnectTimeout)
	} else {
		// Ensure that the ECOM_PG_SSLCERT, ECOM_PG_SSLROOTCERT and ECOM_PG_SSLKEY are all
		// referenced using absolute paths.
		if pgsslcert == "" {
			log.Fatal("missing PostgreSQL SSL certificate file. Use export ECOM_PG_SSLCERT")
		}
		if !filepath.IsAbs(pgsslcert) {
			log.Fatalf("ECOM_PG_SSLCERT should use an absolute path to certificate file to avoid ambiguity")
		}
		mustHaveFile(pgsslcert, "client certificate file")

		if pgsslrootcert == "" {
			log.Fatal("missing PostgreSQL SSL root certificate file. Use export ECOM_PG_SSLROOTCERT")
		}
		if !filepath.IsAbs(pgsslrootcert) {
			log.Fatalf("ECOM_PG_SSLROOTCERT should use an absolute path to root certificate file to avoid ambiguity")
		}
		mustHaveFile(pgsslrootcert, "ssl root certificate")

		if pgsslkey == "" {
			log.Fatal("missing PostgreSQL SSL key certificate file. Use export ECOM_PG_SSLKEY")
		}
		if !filepath.IsAbs(pgsslkey) {
			log.Fatalf("ECOM_PG_SSLKEY should use an absolute path to key certificate file to avoid ambiguity")
		}
		mustHaveFile(pgsslkey, "ssl key file")

		dsn = fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=%s sslmode=%s sslcert=%s sslrootcert=%s sslkey=%s connect_timeout=%s", pghost, pgport, pguser, pgpassword, pgdatabase, pgsslmode, pgsslcert, pgsslrootcert, pgsslkey, pgconnectTimeout)
		log.Infof("postgres dsn: host=%s port=%s user=%s password=***** dbname=%s sslmode=%s sslcert=%s sslrootcert=%s sslkey=%s connect_timeout=%s", pghost, pgport, pguser, pgdatabase, pgsslmode, pgsslcert, pgsslrootcert, pgsslkey, pgconnectTimeout)
	}

	// 2. Service Account Credentials
	if credentialsFile == "" {
		log.Fatal("missing service account credentials file. Use export ECOM_GOOGLE_CREDENTIALS=/path/to/your/service-account-file")
	}
	// if the credentialsFile is a relative pathname, make it relative to the secretVolume/sacDir root
	// i.e. /etc/secret-volume/service_account_credentials/<file>
	if !filepath.IsAbs(credentialsFile) {
		log.Debugf("credentialsFile is a relative pathname so building absolute pathname")
		credentialsFile = filepath.Join(secretVolume, sacDir, credentialsFile)
	}
	mustHaveFile(credentialsFile, "service account credentials file")

	// 3. Google Project ID
	if projectID == "" {
		log.Fatal("missing project ID. Use export ECOM_GOOGLE_PROJECT_ID")
	}
	log.Infof("project ID set to %s", projectID)

	// 4. Server Port
	if port == "" {
		port = "8080"
		log.Infof("no application port specified using default port %s", port)
	} else {
		log.Infof("environment variable ECOM_APP_PORT specifies port %s to be used", port)
	}

	// ensure that we have access to the secret volume
	// Google Kubernetes Engine attaches secrets to /etc/secret-volume mount points
	// Google Compute Engine attaches a persistent disk containing the necessary assets
	// Assets include PostgreSQL .pem files, Google Firebase service account keys and
	// TLS/SSL certificate files for HTTPS termination (see ECOM_APP_TLS_MODE=enable).
	ex, err := exists(secretVolume)
	if err != nil {
		log.Fatalf("failed to determine if secret volume %s exists: %v", secretVolume, err)
	}
	if !ex {
		log.Fatalf("cannot find secret volume %s. Have you mounted it?", secretVolume)
	}
	log.Infof("found secret volume %s", secretVolume)

	// TLS Mode defaults to false unless ECOM_APP_TLS_MODE is set to enable
	// tlsMode will be used to determine whether to provide negociation for SSL
	// (see the bottom of main func)
	tlsMode := false
	if tlsModeFlag == "enable" || tlsModeFlag == "enabled" {
		tlsMode = true
		log.Info("ECOM_APP_TLS_MODE enabled")

		// Ensure the TLS Certificate and Key files exist
		if tlsCertFile == "" {
			log.Fatal("ECOM_APP_TLS_MODE is enabled so you must set the cert file. Use export ECOM_APP_TLS_CERT=/path/to/your/cert.pem")
		}

		// if the tlsCertFile is a relative pathname, make it relative to the secretVolume root
		if !filepath.IsAbs(tlsCertFile) {
			log.Debugf("tlsCertFile is a relative pathname so building absolute pathname")
			tlsCertFile = filepath.Join(secretVolume, tlsCertFile)
		}
		mustHaveFile(tlsCertFile, "TLS Cert File")

		if tlsKeyFile == "" {
			log.Fatal("ECOM_APP_TLS_MODE is enabled so you must set the key file. Use export ECOM_APP_TLS_KEY=/path/to/your/key.pem")
		}
		if !filepath.IsAbs(tlsKeyFile) {
			log.Debugf("tlsKeyFile is a relative pathname so building absolute pathname")
			tlsKeyFile = filepath.Join(secretVolume, tlsKeyFile)
		}
		mustHaveFile(tlsKeyFile, "TLS Key File")
	}

	// connect to postgres
	db, err := sql.Open("postgres", dsn)
	if err != nil {
		log.Fatalf("failed to open db: %v", err)
	}

	attempt := 0
	for attempt < maxDbConnectAttempts {
		err = db.Ping()
		if err != nil {
			attempt++
			if attempt >= maxDbConnectAttempts {
				log.Fatalf("attempt %d/%d failed to verify db connection: %v", attempt, maxDbConnectAttempts, err)
			}
			log.Warnf("attempt %d/%d, failed to verify db connection: %v", attempt, maxDbConnectAttempts, err)
			time.Sleep(5 * time.Second)
		} else {
			break
		}
	}

	log.Infoln("established database connection")

	// build a Postgres model
	pgModel := model.NewPgModel(db)

	// PubSub
	ctx := context.Background()
	opt := option.WithCredentialsFile(credentialsFile)
	log.Infof("initializing Google PubSub client for project %s", projectID)
	psClient, err := pubsub.NewClient(ctx, projectID, opt)
	defer psClient.Close()
	if err != nil {
		log.Fatal(err)
	}

	topic := psClient.Topic(pubsubTopic)
	defer topic.Stop()
	log.Infof("created pubsub topic %v", topic.String())

	// Broadcast startup message
	pr := topic.Publish(ctx, &pubsub.Message{
		Data: []byte("hello world"),
	})

	serverID, err := pr.Get(ctx)

	fmt.Println(serverID, err)

	// build a Google Firebase App
	var fbApp *firebase.App
	var fbAuthClient *auth.Client
	fbApp, err = firebase.NewApp(ctx, nil, opt)
	if err != nil {
		log.Fatalf("%v", fmt.Errorf("failed to initialise Firebase app: %v", err))
	}

	fbAuthClient, err = fbApp.Auth(ctx)
	log.Info("firebase auth initialised")

	// build a Firebase service injecting in the model and firebase app as dependencies
	fbSrv, _ := service.New(pgModel, fbApp, fbAuthClient, psClient)
	a := app.App{
		Service: fbSrv,
	}

	r := chi.NewRouter()

	// protected routes
	r.Group(func(r chi.Router) {
		c := cors.New(cors.Options{
			AllowedOrigins: []string{"*"},
			AllowedHeaders: []string{"Authorization", "Content-Type"},
			AllowedMethods: []string{"GET", "POST", "PATCH", "DELETE"},
			//AllowCredentials: true,
			// Enable Debugging for testing, consider disabling in production
			Debug: false,
		})
		r.Use(c.Handler)
		r.Use(a.AuthenticateMiddleware)

		// Customer and address management API
		r.Route("/customers", func(r chi.Router) {
			r.Get("/", a.Authorization("ListCustomers", a.ListCustomersHandler()))
			r.Post("/", a.Authorization("CreateCustomer", a.CreateCustomerHandler()))
			r.Get("/{cuuid}", a.Authorization("GetCustomer", a.GetCustomerHandler()))
			r.Post("/{cuuid}/addresses", a.Authorization("CreateAddress", a.CreateAddressHandler()))
			r.Get("/{cuuid}/addresses", a.Authorization("GetCustomersAddresses", a.ListAddressesHandler()))
			r.Patch("/{cuuid}/addresses/{auuid}", a.Authorization("UpdateAddress", a.UpdateAddressHandler()))
		})

		r.Route("/addresses", func(r chi.Router) {
			r.Get("/{auuid}", a.Authorization("GetAddress", a.GetAddressHandler()))
			r.Delete("/{auuid}", a.Authorization("DeleteAddress", a.DeleteAddressHandler()))
		})

		r.Route("/carts", func(r chi.Router) {
			r.Post("/", a.Authorization("CreateCart", a.CreateCartHandler()))
			r.Post("/{ctid}/items", a.Authorization("AddItemToCart", a.AddItemToCartHandler()))
			r.Get("/{ctid}/items", a.Authorization("GetCartItems", a.GetCartItemsHandler()))
			r.Patch("/{ctid}/items/{sku}", a.Authorization("UpdateCartItem", a.UpdateCartItemHandler()))
			r.Delete("/{ctid}/items/{sku}", a.Authorization("DeleteCartItem", a.DeleteCartItemHandler()))
			r.Delete("/{ctid}/items", a.Authorization("EmptyCartItems", a.EmptyCartItemsHandler()))
		})

		r.Route("/catalog", func(r chi.Router) {
			r.Get("/", a.Authorization("GetCatalog", a.GetCatalogHandler()))
		})
		r.Get("/productassocs", a.GetCatalogProductAssocsHandler())
	})

	// public routes including GET / for Google Kuberenetes default healthcheck
	r.Group(func(r chi.Router) {
		// r.Use(middleware.NoCache)

		// version info
		r.Get("/", healthCheckHandler)
		r.Get("/healthz", healthCheckHandler)
		r.Get("/info", infoHandler)
	})

	// tlsMode determines whether to serve HTTPS traffic directly.
	// If tlsMode is false, you can enable HTTPS with a GKE Layer 7 load balancer
	// using an Ingress.
	if tlsMode {
		log.Infof("server listening on HTTPS port %s", port)
		log.Fatal(http.ListenAndServeTLS(fmt.Sprintf(":%s", port), tlsCertFile, tlsKeyFile, r))
	} else {
		log.Infof("server listening on HTTP port %s", port)
		log.Fatal(http.ListenAndServe(fmt.Sprintf(":%s", port), r))
	}
}
